### PR TITLE
02-client routing: implement `VerifyUpgradeAndUpdateState` in light client modules

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

closes: #XXXX

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/keeper/msg_server.go [hunk 1]
```diff
@@ -69,17 +69,13 @@ func (k Keeper) UpdateClient(goCtx context.Context, msg *clienttypes.MsgUpdateCl
...
-	if err != nil {
-		return nil, err
-	}
-
-	if err = k.ClientKeeper.UpgradeClient(ctx, msg.ClientId, upgradedClient, upgradedConsState,
-		msg.ProofUpgradeClient, msg.ProofUpgradeConsensusState); err != nil {
+	if err := k.ClientKeeper.UpgradeClient(
+		ctx, msg.ClientId,
+		msg.ClientState.Value,
+		msg.ConsensusState.Value,
```
Comment (crodriguezvega) [inline modules/core/keeper/msg_server.go:75]: @colin-axner @damiannolan I think just passing doing `.Value` here and passing the bytes of the proton-encoded client and consensus states might not be enough. See test failing [here](https://github.com/cosmos/ibc-go/actions/runs/7849754768/job/21423773040?pr=5827), it looks like if we just pass the bytes we miss the type URL information that we need to unmarshal the bytes.

File: modules/light-clients/07-tendermint/light_client_module.go [hunk 2]
```diff
@@ -295,30 +303,45 @@ func (lcm LightClientModule) VerifyUpgradeAndUpdateState(
...
+
+	if err := cdc.UnmarshalInterface(newClient, &newClientState); err != nil {
 		return err
 	}
-
+	newTmClientState, ok := newClientState.(*ClientState)
+	if !ok {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidClient, "expected client state type %T, got %T", (*ClientState)(nil), newClientState)
+	}
 	if err := newClientState.Validate(); err != nil {
```
Comment (crodriguezvega) [inline modules/light-clients/07-tendermint/light_client_module.go:319]: @colin-axner @damiannolan This validation fails if the trust level is not within [0, 1/3] (see test failing [here](https://github.com/cosmos/ibc-go/actions/runs/7849754768/job/21423772705?pr=5827)). However, the upgraded client state that the counterparty chain commits should have all the custom fields zeroed out [since that's what we verify in `VerifyUpgradeAndUpdateState` function in the `ClientState`](https://github.com/cosmos/ibc-go/blob/d957ef80d72a260af3f6e66bbf5d7bea85473e2b/modules/light-clients/07-tendermint/upgrade.go#L73). So I think we either cannot do this validation here or we need to validate only the fields that shouldn't be zeroed out.

File: modules/core/keeper/msg_server.go (hunk 1 above)
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:75]: See https://github.com/cosmos/ibc-go/pull/5827/commits/3e311639ef002c2272ed02e11cbd5aee1bc543d3

We should unmarshal to concrete types and not interfaces within the light client module.

File: modules/light-clients/07-tendermint/light_client_module.go (hunk 2 above)
Comment (damiannolan) [reply modules/light-clients/07-tendermint/light_client_module.go]: I saw this failing test when working on the one above. I didn't dig any deeper on this one yet. But from what you're saying here it sounds correct. I'm not sure exactly what pieces of code we'll need to change yet, can follow up o this one next week! :)

File: modules/core/02-client/keeper/client.go [hunk 3]
```diff
@@ -122,46 +121,41 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, clientMsg exporte
 
 // UpgradeClient upgrades the client to a new client state if this new client was committed to
 // by the old client at the specified upgrade height
-func (k Keeper) UpgradeClient(ctx sdk.Context, clientID string, upgradedClient exported.ClientState, upgradedConsState exported.ConsensusState,
-	upgradeClientProof, upgradeConsensusStateProof []byte,
+func (k Keeper) UpgradeClient(
+	ctx sdk.Context,
+	clientID string,
+	upgradedClient, upgradedConsState, upgradeClientProof, upgradeConsensusStateProof []byte,
```
Comment (damiannolan) [inline modules/core/02-client/keeper/client.go:127]: Isn't this an API breaking change? shouldn't the API remain the same here in the 02-client keeper and continue to pass the `exported` interfaces.

We can then pass `.Value` bytes when calling into the light client module.

File: modules/core/keeper/msg_server.go (hunk 1 above)
Comment (damiannolan) [inline modules/core/keeper/msg_server.go:75]: essentially push this down a level into 02-client

File: modules/core/keeper/msg_server_test.go [hunk 4]
```diff
@@ -888,7 +888,7 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
 		tc.setup()
 
 		ctx := suite.chainA.GetContext()
-		_, err = keeper.Keeper.UpgradeClient(*suite.chainA.App.GetIBCKeeper(), ctx, msg)
+		_, err = suite.chainA.GetSimApp().GetIBCKeeper().UpgradeClient(ctx, msg)
```
Comment (damiannolan) [inline modules/core/keeper/msg_server_test.go:891]: decided to change this while I was here, makes no difference but kinda looks weird how it was 🧹 

I think these tests could be improved or cleaned up some more. But maybe outside the scope of this PR

File: modules/core/02-client/keeper/client.go (hunk 3 above)
Comment (crodriguezvega) [reply modules/core/02-client/keeper/client.go:127]: It is API breaking, but I think that this refactoring will be API breaking anyway and will be released in a major release. I basically did here [a similar change as was done in `CreateClient`](https://github.com/cosmos/ibc-go/blob/f4ee20bf8c9139c7806c6410785ae37073bc7702/modules/core/02-client/keeper/client.go#L20).

File: modules/core/keeper/msg_server.go (hunk 1 above)
Comment (crodriguezvega) [reply modules/core/keeper/msg_server.go:75]: Thanks, Damian. I changed this to unmarshal to interface because [the tests are generally using `client types.MarshalClientState`](https://github.com/cosmos/ibc-go/blob/055a3e39c3da309e013ec4763f78d33b74be59c9/modules/core/02-client/keeper/client_test.go#L481), which [calls `cdc.MarshalInterface`](https://github.com/cosmos/ibc-go/blob/055a3e39c3da309e013ec4763f78d33b74be59c9/modules/core/02-client/types/encoding.go#L35), so when trying to unmarshal to the concrete type I would get [errors like the ones you see now in the PR](https://github.com/cosmos/ibc-go/actions/runs/7952186466/job/21706474023?pr=5827).

So I guess my question would be: do we expect the client state (and consensus state) received in the messages be marshalled as an interface or a concrete type? I am not sure if this question really makes sense, though. Maybe relayers will just send the bytes for the marshalled concrete type (it seems like [that's what hermes does](https://github.com/informalsystems/hermes/blob/f986dc82cff5aafef11befee44b1d2eb8d179ad9/crates/relayer/src/client_state.rs#L149-L152), although the [go relayer seems to marshal the interface type](https://github.com/cosmos/relayer/blob/11b7de318b5527b0f097475df1865fe472f76ff1/relayer/chains/cosmos/tx.go#L708-L716)?), so maybe we just need to change our tests to also marshal to the concrete type? Sorry, my protobuf foo is pretty weak, so I might just not making any sense, hehe. But having a bit of clarity on this would be helpful for me to unblock this work.

File: modules/light-clients/07-tendermint/light_client_module.go (hunk 2 above)
Comment (damiannolan) [reply modules/light-clients/07-tendermint/light_client_module.go]: I removed the basic validation of client state and consensus state at the light client module level. ClientState is reconstructed within the method `VerifyUpgradeAndUpdateState` and validated there. 

There is still some test failures wrt encoding from what I can see, but I think they should be manageable. Is the plan to restructure tests to invoke the client module in this PR or to leave it for a follow up? 

Happy to help out where needed :)

File: modules/light-clients/08-wasm/light_client_module.go [hunk 5]
```diff
@@ -276,14 +277,69 @@ func (lcm LightClientModule) RecoverClient(ctx sdk.Context, clientID, substitute
...
-// // DEPRECATED: will be removed as performs internal functionality
-func (LightClientModule) VerifyUpgradeAndUpdateState(ctx sdk.Context, clientID string, newClient []byte, newConsState []byte, upgradeClientProof, upgradeConsensusStateProof []byte) error {
-	return nil
+// Upgrade functions
+// NOTE: proof heights are not included as upgrade to a new revision is expected to pass only on the last
+// height committed by the current revision. Clients are responsible for ensuring that the planned last
+// height of the current revision is somehow encoded in the proof verification process.
+// This is to ensure that no premature upgrades occur, since upgrade plans committed to by the counterparty
+// may be cancelled or modified before the last planned height.
+// If the upgrade is verified, the upgraded client and consensus states must be set in the client store.
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/light_client_module.go:286]: godoc should probably be rewritten to be wasm specific, not a copy/paste of the API godoc

File: modules/light-clients/08-wasm/light_client_module.go [hunk 6]
```diff
@@ -276,14 +277,69 @@ func (lcm LightClientModule) RecoverClient(ctx sdk.Context, clientID, substitute
...
+		return errorsmod.Wrapf(clienttypes.ErrInvalidClientType, "expected: %s, got: %s", types.Wasm, clientType)
+	}
+
+	var (
+		cdc               = lcm.keeper.Codec()
+		newClientState    exported.ClientState
+		newConsensusState exported.ConsensusState
+	)
+
+	if err := cdc.UnmarshalInterface(newClient, &newClientState); err != nil {
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/light_client_module.go:310]: I wonder if we could unmarshal directly into the concrete type and drop the cast assertions that follow?

File: modules/light-clients/07-tendermint/light_client_module_test.go [hunk 7]
```diff
@@ -137,3 +144,173 @@ func (suite *TendermintTestSuite) TestRecoverClient() {
...
+			ibcerrors.ErrInvalidHeight,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+			cdc := suite.chainA.App.AppCodec()
+			ctx := suite.chainA.GetContext()
```
Comment (colin-axner) [inline modules/light-clients/07-tendermint/light_client_module_test.go:265]: super nit: one thing I dislike about assigning the context to a variable rather than dynamically requesting it is that this variable has the incorrect block height set in the context. The `tc.malleate()` call will commit a block on chainA on the "success" case, resulting in the context having an out of date block header when calling `VerifyUpgradeAndUpdateState`. I guess it is fine in this test, but can lead to confusing errors if it needs to be accurate

Comment (crodriguezvega) [reply modules/light-clients/07-tendermint/light_client_module_test.go]: Got it. I changed it now.

File: modules/light-clients/08-wasm/light_client_module.go (hunk 5 above)
Comment (crodriguezvega) [reply modules/light-clients/08-wasm/light_client_module.go]: Fixed.

File: modules/core/02-client/keeper/client_test.go [hunk 8]
```diff
@@ -351,64 +351,35 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
...
 
-				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(lastHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
-				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(lastHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
+				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(upgradeHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
+				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(upgradeHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
 			},
 			expPass: true,
 		},
-		{
-			name: "client state not found",
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/client_test.go:360]: I deleted this test because this is tested in the light client module.

File: modules/core/02-client/keeper/client_test.go [hunk 9]
```diff
@@ -440,71 +412,44 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
 				cs, found := suite.chainA.App.GetIBCKeeper().ClientKeeper.GetClientState(suite.chainA.GetContext(), path.EndpointA.ClientID)
 				suite.Require().True(found)
 
-				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(lastHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
-				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(lastHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
-			},
-			expPass: false,
-		},
-		{
-			name: "unsuccessful upgrade: upgraded height is not greater than current height",
```
Comment (crodriguezvega) [inline modules/core/02-client/keeper/client_test.go:449]: I moved this test to the light client module, but can move it back here in the PR where `LatestHeight` is added to the light client module and then we can move the check back to 02-client.

File: modules/light-clients/08-wasm/light_client_module.go (hunk 6 above)
Comment (crodriguezvega) [reply modules/light-clients/08-wasm/light_client_module.go:263]: As discussed on Monday, we need to unmarshal into the interface.

**Comment by crodriguezvega** [comment]
I think this is again ready for review after the discussion on Monday. There are a couple of tests that fail because we need to add the right value for the latest height in the create and upgrade client events, but that should be fixed in https://github.com/cosmos/ibc-go/pull/5866.

File: modules/core/keeper/msg_server.go [hunk 10]
```diff
@@ -69,17 +69,23 @@ func (k Keeper) UpdateClient(goCtx context.Context, msg *clienttypes.MsgUpdateCl
 func (k Keeper) UpgradeClient(goCtx context.Context, msg *clienttypes.MsgUpgradeClient) (*clienttypes.MsgUpgradeClientResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	upgradedClient, err := clienttypes.UnpackClientState(msg.ClientState)
+	upgradedClientState, err := k.cdc.Marshal(msg.ClientState)
```
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:72]:
```suggestion
        // Backwards Compatibility: the light client module is expecting
        // to unmarshal an interface so we marshal the client state Any
	upgradedClientState, err := k.cdc.Marshal(msg.ClientState)
```

File: modules/core/keeper/msg_server.go [hunk 11]
```diff
@@ -69,17 +69,23 @@ func (k Keeper) UpdateClient(goCtx context.Context, msg *clienttypes.MsgUpdateCl
...
 	if err != nil {
 		return nil, err
 	}
-	upgradedConsState, err := clienttypes.UnpackConsensusState(msg.ConsensusState)
+
+	upgradedConsensusState, err := k.cdc.Marshal(msg.ConsensusState)
```
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:77]:
```suggestion
        // Backwards Compatibility: the light client module is expecting
        // to unmarshal an interface so we marshal the consensus state Any
	upgradedConsensusState, err := k.cdc.Marshal(msg.ConsensusState)
```

some sort of in line code doc like this can be useful

File: modules/light-clients/07-tendermint/light_client_module.go [hunk 12]
```diff
@@ -256,30 +266,39 @@ func (lcm LightClientModule) VerifyUpgradeAndUpdateState(
...
 	if !found {
 		return errorsmod.Wrap(clienttypes.ErrClientNotFound, clientID)
 	}
 
-	return clientState.VerifyUpgradeAndUpdateState(ctx, cdc, clientStore, &newClientState, &newConsensusState, upgradeClientProof, upgradeConsensusStateProof)
+	// last height of current counterparty chain must be client's latest height
+	lastHeight := clientState.GetLatestHeight()
+	if !newClientState.GetLatestHeight().GT(lastHeight) {
+		return errorsmod.Wrapf(ibcerrors.ErrInvalidHeight, "upgraded client height %s must be at greater than current client height %s", newClientState.GetLatestHeight(), lastHeight)
+	}
```
Comment (colin-axner) [inline modules/light-clients/07-tendermint/light_client_module.go:301]: Add back to 02-client level once latest height pr is merged?

**Review by colin-axner — APPROVED** [review]
Thanks @crodriguezvega!!

**Comment by damiannolan** [comment]
Resolving conflicts.. 🏃🏻

File: modules/core/02-client/keeper/client.go [hunk 13]
```diff
@@ -122,46 +121,41 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, clientMsg exporte
...
+		return errorsmod.Wrapf(types.ErrClientNotFound, "clientID (%s)", clientID)
+	}
 
-	if !upgradedClient.GetLatestHeight().GT(lastHeight) {
-		return errorsmod.Wrapf(ibcerrors.ErrInvalidHeight, "upgraded client height %s must be at greater than current client height %s",
-			upgradedClient.GetLatestHeight(), lastHeight)
+	lightClientModule, found := k.router.GetRoute(clientID)
+	if !found {
+		return errorsmod.Wrap(types.ErrRouteNotFound, clientID)
 	}
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/02-client/keeper/client.go:142]: discussed async, this check must be moved to the light client module, and this functionality should likely only exist in the tendermint client module.

At first glance, we do not have the `clientID` of the `upgradedClientState`. However, it should be the same `clientID` as the `clientState` (target). However, we need the latest height from the `upgradedClientState` concrete implementation here - this is the client which the counterparty stores in it upgrade path and we prove against within the tm client module.

File: modules/core/02-client/keeper/client_test.go [hunk 14]
```diff
@@ -448,73 +448,73 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
...
-				suite.Require().True(ok)
-
-				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(lastHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
-				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(lastHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
+				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(upgradeHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
+				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(upgradeHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
 			},
 			expPass: false,
 		},
+		// TODO: Needs to move to light client module?
```
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:456]: this test needs to move given that we have removed the height check

File: modules/core/02-client/keeper/client_test.go (hunk 9 above)
Comment (damiannolan) [reply modules/core/02-client/keeper/client_test.go:457]: https://github.com/cosmos/ibc-go/pull/5827#discussion_r1520067044

I guess it can just be removed entirely now?

File: modules/core/keeper/msg_server.go [hunk 15]
```diff
@@ -70,17 +70,26 @@ func (k Keeper) UpdateClient(goCtx context.Context, msg *clienttypes.MsgUpdateCl
 func (k Keeper) UpgradeClient(goCtx context.Context, msg *clienttypes.MsgUpgradeClient) (*clienttypes.MsgUpgradeClientResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	upgradedClient, err := clienttypes.UnpackClientState(msg.ClientState)
+	// Backwards Compatibility: the light client module is expecting
+	// to unmarshal an interface so we marshal the client state Any
+	upgradedClientState, err := k.cdc.Marshal(msg.ClientState)
```
Comment (AdityaSripal) [inline modules/core/keeper/msg_server.go:75]: It's a bit unfortunate that we're marshalling here and then unmarshalling later in the callstack. Seems wasteful...

**Review by AdityaSripal — COMMENTED** [review]
Looks good! I suppose we have to keep the interfaces for users the same way, but would be nice to eventually just pass everything through and unmarshal at the light client module level.

File: modules/core/02-client/keeper/client_test.go (hunk 9 above)
Comment (crodriguezvega) [reply modules/core/02-client/keeper/client_test.go:457]: Sounds good. At the moment this check is performed in 07-tendermint and 08-wasm light client modules. Based on [your comment](https://github.com/cosmos/ibc-go/pull/5827/files#r1520066359), is that ok? Any reason why the check should not be done in 08-wasm?

File: modules/core/keeper/msg_server.go (hunk 15 above)
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:75]: I don't think this is necessary. I think this is symptom of how the tests were/are written and they just need to be refactored. After looking closer at the code and how it works, I think it should be possible to pass the `Any.Value` bytes along here and into the light client module and have it work.

Opened a draft which I'll pick up in the morning, but see here for happy path test working: https://github.com/cosmos/ibc-go/pull/5967/files#diff-cef346df04f3fb2455741328504cc2cbb06c7133cce6ede300347139fbdda74eR490

tldr; test code needs to define `*codectypes.Any` and pass their value bytes as args. We can then make the changes in the light client module to unmarshal to concrete types. When we reconstruct the clientState which is proven it has always used `MarshalInterface` given the concrete type and should continue to work like that. I think the tests just need more attention.

File: modules/core/02-client/keeper/client_test.go (hunk 9 above)
Comment (charleenfei) [reply modules/core/02-client/keeper/client_test.go:457]: we can have these tests in the lcm modules for 07-tendermint and 08-wasm already, i think! is that fine @crodriguezvega?

Comment (crodriguezvega) [reply modules/core/02-client/keeper/client_test.go:457]: Yes, correct! The check and the tests are currently in the light client module. Happy to leave it there if that's the right place. :)

File: modules/core/02-client/keeper/client_test.go [hunk 16]
```diff
@@ -353,21 +354,21 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
...
-				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(lastHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
+				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(upgradeHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
+				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(upgradeHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
 			},
 			expPass: true,
 		},
 		{
 			name: "client state not found",
 			setup: func() {
 				// last Height is at next block
```
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:365]:
```suggestion
				// upgrade height is at next block
```

File: modules/core/02-client/keeper/client_test.go [hunk 17]
```diff
@@ -394,16 +395,15 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
 				// client is frozen
 
 				// last Height is at next block
```
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:397]:
```suggestion
				// upgrade height is at next block
```

File: modules/core/02-client/keeper/client_test.go [hunk 18]
```diff
@@ -425,49 +425,22 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
 			expPass: false,
 		},
 		{
-			name: "tendermint client VerifyUpgrade fails",
+			name: "light client module VerifyUpgradeAndUpdateState fails",
 			setup: func() {
 				// last Height is at next block
```
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:430]:
```suggestion
				// upgrade height is at next block
```

File: modules/light-clients/08-wasm/light_client_module.go [hunk 19]
```diff
@@ -257,6 +258,53 @@ func (lcm LightClientModule) RecoverClient(ctx sdk.Context, clientID, substitute
...
+	if err := cdc.UnmarshalInterface(newConsState, &newConsensusState); err != nil {
+		return err
+	}
+	newWasmConsensusState, ok := newConsensusState.(*types.ConsensusState)
+	if !ok {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidConsensus, "expected consensus state type %T, got %T", (*types.ConsensusState)(nil), newConsensusState)
+	}
+	if err := newWasmConsensusState.ValidateBasic(); err != nil {
+		return err
+	}
```
Comment (damiannolan) [inline modules/light-clients/08-wasm/light_client_module.go:295]: This code needs to be aligned with 07-tendermint impl, shame that this is not caught with tests.. hmmm

File: modules/light-clients/08-wasm/light_client_module_test.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/light_client_module_test.go:1]: And tests

**Review by damiannolan — CHANGES_REQUESTED** [review]
I missed 08-wasm changes needed to aligned with 07-tendermint from my encoding PR. 
I'll handle them today and then LGTM! 🙏🏻

File: modules/core/02-client/keeper/client_test.go [hunk 20]
```diff
@@ -381,8 +381,8 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
 				tmCs, ok := cs.(*ibctm.ClientState)
 				suite.Require().True(ok)
 
-				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(lastHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
-				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(lastHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
+				upgradedClientProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(upgradeHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
+				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(upgradeHeight.GetRevisionHeight())), tmCs.LatestHeight.GetRevisionHeight())
 
 				path.EndpointA.ClientID = "wrongclientid"
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/core/02-client/keeper/client_test.go:388]: can follow up this post merge, maybe? I think all these tests should be using `expError` and not `expPass` (bool)

**Review by damiannolan — APPROVED** [review]
LGTM!
